### PR TITLE
Make architecture more informative

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

File: docs/dev/architecture.md [hunk 1]
```diff
@@ -1,174 +1,375 @@
...
 
-Although `hir` and `ide` don't do any IO, we need to be able to read
-files from disk at the end of the day. This is what `vfs` does. It also
-manages overlays: "dirty" files in the editor, whose "true" contents is
-different from data on disk. 
+They use `crates/path` heavily instead of `std::path`.
+A single `rust-analyzer` process can serve many projects, so it is important that server's current directory does not leak.
 
-## Testing Infrastructure
+### `crates/mbe`, `crated/tt`, `crates/proc_macro_api`, `crates/proc_macro_srv`
```
Comment (matklad) [inline docs/dev/architecture.md:252]: @edwin0cheng if you can expand this section, that't be awesome! It's worth explaining the difference between srv and api, relation to the code in rustc and self-execution trick we use.

**Comment by matklad** [comment]
@Veykril you are a relatively fresh face and you've touched many of the moving pieces, so feel free to submit follow-ups if you feel something's missing/unclear!

File: docs/dev/architecture.md (hunk 1 above)
Comment (edwin0cheng) [reply docs/dev/architecture.md:252]: Sure, will do !

Comment (edwin0cheng) [reply docs/dev/architecture.md:252]: #7584

File: docs/dev/architecture.md [hunk 2]
```diff
@@ -1,174 +1,375 @@
...
+It works similarly to
+[kotlin's parser](https://github.com/JetBrains/kotlin/blob/4d951de616b20feca92f3e9cc9679b2de9e65195/compiler/frontend/src/org/jetbrains/kotlin/parsing/KotlinParsing.java),
+which is a good source of inspiration for dealing with syntax errors and incomplete input.
+Original [libsyntax parser](https://github.com/rust-lang/rust/blob/6b99adeb11313197f409b4f7c4083c2ceca8a4fe/src/libsyntax/parse/parser.rs) is what we use for the definition of the Rust language.
+`TreeSink` and `TokenSource` traits bridge the tree-agnostic parser from `grammar` with `rowan` trees.
 
-### `crates/ra_syntax`, `crates/parser`
+**Architecture Invariant:** the parser is independent of the particular tree structure and particular representation of the tokens.
+It transforms one flat stream of events into another flat stream of events.
+Token independence allows us to pares out both text-based source code and `tt`-based macro input.
```
Comment (bjorn3) [inline docs/dev/architecture.md:78]: *parse

File: docs/dev/architecture.md [hunk 3]
```diff
@@ -1,174 +1,375 @@
...
+
+To update test data, run with `UPDATE_EXPECT` variable:
 
-Tests for ra_syntax are mostly data-driven: `test_data/parser` contains subdirectories with a bunch of `.rs`
-(test vectors) and `.txt` files with corresponding syntax trees. During testing, we check
-`.rs` against `.txt`. If the `.txt` file is missing, it is created (this is how you update
-tests). Additionally, running `cargo xtask codegen` will walk the grammar module and collect
-all `// test test_name` comments into files inside `test_data/parser/inline` directory.
+```bash
+env UPDATE_EXPECT=1 cargo qt
```
Comment (bjorn3) [inline docs/dev/architecture.md:103]: `cargo test`?

**Pull request merged**